File: cpp/src/ch08/maze_state.cc
#include <deque>
#include <sstream>
#include <memory>
#include "maze_state.h"

bool ConnectFourState::is_done() const
{
    return win_status_ != GameStatus::ONGOING;
}

// start ignore "-Wreturn-type"
// this doesn't return Stone on all paths but we know
// the break condition is always met
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wreturn-type"

Stone ConnectFourStateNormal::place_stone(const int action)
{
    // get first stone to place
    for (int y = 0; y < H; y++)
    {
        if (my_board_[y][action] == 0 && enemy_board_[y][action] == 0)
        {
            my_board_[y][action] = 1;
            return Stone(y, action);
        }
    }
}

#pragma GCC diagnostic pop // end ignore "-Wreturn-type"

void ConnectFourStateNormal::check_connection(const Stone first_stone,
                                              const int dx[2],
                                              const int dy[2])
{
    auto que = std::deque<Stone>();
    que.emplace_back(first_stone);
    std::vector<std::vector<bool>> check(H, std::vector<bool>(W, false));
    int count = 0;

    while (!que.empty())
    {
        const auto &stone = que.front();
        que.pop_front();
        count++;
        if (count >= 4)
        {
            // 自分が揃ったら相手視点は負け
            // is_first の チェックが不要なのかが疑問
            // is_done を踏む瞬間は自分が必ず負けているのでOK
            win_status_ = GameStatus::LOSE;
            return;
        }
        check[stone.y_][stone.x_] = true;

        for (int action = 0; action < 2; action++)
        {
            int ty = stone.y_ + dy[action];
            int tx = stone.x_ + dx[action];

            if (ty >= 0 && ty < H && tx >= 0 && tx < W &&
                my_board_[ty][tx] == 1 && !check[ty][tx])
            {
                que.emplace_back(ty, tx);
            }
        }
    }
}

std::string ConnectFourStateNormal::to_string() const
{
    std::stringstream ss("");
    ss << "\n";
    ss << "is_first: " << is_first_ << "\n";
    for (int y = H - 1; y >= 0; y--)
    {
        ss << "\n";
        for (int x = 0; x < W; x++)
        {
            char c = '.';
            if (my_board_[y][x] == 1)
                c = (is_first_ ? 'X' : 'O');
            else if (enemy_board_[y][x] == 1)
                c = (is_first_ ? 'O' : 'X');
            ss << c;
        } // x
    }     // y

    ss << "\n";
    return ss.str();
}

std::string ConnectFourStateBitset::to_string() const
{
    std::stringstream ss("");
    ss << "\n";
    ss << "is_first: " << is_first_ << "\n";
    for (int y = H - 1; y >= 0; y--)
    {
        ss << "\n";
        for (int x = 0; x < W; x++)
        {
            int index = x * (H + 1) + y;
            char c = '.';
            if ((my_bit_board_ & (1ULL << index)) != 0)
                c = (is_first_ ? 'X' : 'O');
            else if ((all_bit_board_ & (1ULL << index)) != 0)
                c = (is_first_ ? 'O' : 'X');
            ss << c;
        } // x
    }     // y

    ss << "\n";
    return ss.str();
}

double ConnectFourState::teban_score() const
{
    switch (win_status_)
    {
    case GameStatus::WIN:
        return 1.0;
    case GameStatus::DRAW:
        return 0.5;
    case GameStatus::LOSE:
        return 0.0;
    default:
        // should not reach any other case
        // if not done, will not call teban_score()
        return 0.5;
    }
}

double ConnectFourState::white_score() const
{
    double score = teban_score();
    if (!is_first_)
        score = 1 - score;
    return score;
}

std::vector<int> ConnectFourStateNormal::legal_actions() const
{
    std::vector<int> actions;
    for (int x = 0; x < W; x++)
    {
        for (int y = 0; y < H; y++)
        {
            if (my_board_[y][x] == 0 && enemy_board_[y][x] == 0)
            {
                actions.emplace_back(x);
                break;
            }
        }
    }
    return actions;
}
void ConnectFourStateNormal::advance(const int action)
{
    Stone stone = place_stone(action);

    // dx の増減をチェックすることで横方向の連携判定
    check_connection(stone, d_up, d_stay);

    if (!is_done())
    {
        // "/" 方向のチェックは {1, -1}, {1, -1}
        check_connection(stone, d_up, d_up);
    }

    if (!is_done())
    {
        // "\" 方向のチェックは {1, -1}, {-1, 1}
        check_connection(stone, d_up, d_down);
    }

    if (!is_done())
    {
        // 上下方向（下方向）のcheck
        // 上には石は無いので若干無駄ではあるが、
        // consistency の方が大事なので
        // {0, 0}, {1, -1} をチェック
        check_connection(stone, d_stay, d_up);
    }

    std::swap(my_board_, enemy_board_);
    is_first_ = !is_first_;
    if (!is_done() && legal_actions().size() == 0)
    {
        win_status_ = GameStatus::DRAW;
    }
}

uint64_t ConnectFourStateBitset::floor_bit(int w, int h) const
{
    // 0b00000010000001...
    // w: 繰り返しの数 h: 0 の数
    uint64_t bit = 0;
    for (int x = 0; x < w; x++)
        bit |= 1ULL << x * (h + 1);
    return bit;
}

uint64_t ConnectFourStateBitset::filled(int w, int h) const
{
    /*
    0b01111110111111...
    makes a head missing bit, this means all  board is filled

    1 << h-1 = 100
    100 - 1 = 011
    011 << x*h = 011000
    の組み合わせ
    */
    uint64_t bit = 0;
    for (int x = 0; x < w; x++)
    {
        bit |= ((1ULL << (h - 1)) - 1) << x * h;
    }
    return bit;
}

std::vector<int> ConnectFourStateBitset::legal_actions() const
{
    std::vector<int> actions;
    uint64_t possible = all_bit_board_ + floor_bit(W, H);
    uint64_t filter = 0b111111;
    for (int x = 0; x < W; x++)
    {
        if ((filter & possible) != 0)
            actions.emplace_back(x);
        filter <<= (H + 1);
    }
    return actions;
}

void ConnectFourStateBitset::advance(const int action)
{
    // 敵視点に切り替え (分かりづらいが、all bit board の加算前にやる必要がある)
    my_bit_board_ ^= all_bit_board_;
    is_first_ = !is_first_;

    all_bit_board_ |= (all_bit_board_ + (1ULL << action * (H + 1)));

    // 自分視点に切り替え (無駄なようだがこの順が必要)
    if (is_winner(my_bit_board_ ^ all_bit_board_))
        win_status_ = GameStatus::LOSE;
    else if (all_bit_board_ == filled(W, H))
        win_status_ = GameStatus::DRAW;
}

bool ConnectFourStateBitset::is_winner(const uint64_t board)
{
    // -- dir
    uint64_t tmp_board = board & (board >> 7);
    if ((tmp_board & (tmp_board >> 14)) != 0)
        return true;

    // \ dir
    tmp_board = board & (board >> 6);
    if ((tmp_board & (tmp_board >> 12)) != 0)
        return true;

    // / dir
    tmp_board = board * (board >> 8);
    if ((tmp_board & (tmp_board >> 16)) != 0)
        return true;

    // | dir
    tmp_board = board & (board >> 1);
    if ((tmp_board & (tmp_board >> 2)) != 0)
        return true;

    return false;
}

std::unique_ptr<ConnectFourState> get_state(StateVersion version)
{
    switch (version)
    {
    case StateVersion::Normal:
        return std::make_unique<ConnectFourStateNormal>();
    case StateVersion::Bitset:
        return std::make_unique<ConnectFourStateBitset>();
    }
}

File: rust/src/ch08/maze_state.rs
use crate::base::alternate::{AlternateState, MazeParams};
use std::{collections::VecDeque, mem::swap};

#[derive(Debug, Clone, PartialEq, Eq)]
struct Stone {
    x: usize,
    y: usize,
}

// remake enum in python
enum D {
    UP,
    DOWN,
    STAY,
}

impl D {
    fn value(&self) -> Vec<isize> {
        match self {
            D::STAY => vec![0, 0],
            D::DOWN => vec![-1, 1],
            D::UP => vec![1, -1],
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum Status {
    ONGOING,
    // todo: comment out the below with comments
    // WIN is never made, all advance ends on LOSE or DRAW
    WIN,
    LOSE,
    DRAW,
}

#[derive(Debug, Clone)]
struct ConnectFourState {
    is_first: bool,
    my_board: Vec<Vec<bool>>,
    enemy_board: Vec<Vec<bool>>,
    status: Status,
    params: MazeParams,
}

impl AlternateState for ConnectFourState {
    // seed is not used but necessary for keeping the same interface
    fn new(_: u64, params: MazeParams) -> Self {
        ConnectFourState {
            is_first: true,
            my_board: vec![vec![false; params.width]; params.height],
            enemy_board: vec![vec![false; params.width]; params.height],
            status: Status::ONGOING,
            params: params.clone(),
        }
    }

    fn legal_actions(&self) -> Vec<usize> {
        let mut actions = vec![];
        for x in 0..self.params.width {
            for y in 0..self.params.height {
                if !self.my_board[y][x] && !self.enemy_board[y][x] {
                    actions.push(x);
                    break;
                }
            }
        }
        actions
    }
    fn to_string(&self) -> String {
        let mut ss = format!("is_first: {}\n", self.is_first);
        for y in (0..self.params.height).rev() {
            ss += "\n";
            for x in 0..self.params.width {
                if self.my_board[y][x] {
                    ss.push_str(if self.is_first { "X" } else { "O" });
                } else if self.enemy_board[y][x] {
                    ss.push_str(if self.is_first { "O" } else { "X" });
                } else {
                    ss += ".";
                }
            }
        }
        ss += "\n";
        ss
    }

    fn teban_score(&self) -> f32 {
        match self.status {
            Status::DRAW => 0.5,
            Status::LOSE => 0.0,
            Status::WIN => 1.0,
            Status::ONGOING => panic!("shouldn't call teban_score"),
        }
    }

    fn white_score(&self) -> f32 {
        let score = self.teban_score();
        if !self.is_first {
            return 1.0 - score;
        }
        score
    }
    fn is_done(&self) -> bool {
        self.status != Status::ONGOING
    }

    fn advance(&mut self, action: usize) {
        let stone = self.place_stone(action);

        // -
        self.check_connection(&stone, D::UP, D::STAY);
        // /
        if !self.is_done() {
            self.check_connection(&stone, D::UP, D::UP);
        }
        // \
        if !self.is_done() {
            self.check_connection(&stone, D::UP, D::DOWN);
        }
        // |
        if !self.is_done() {
            self.check_connection(&stone, D::STAY, D::UP);
        }

        // swap
        swap(&mut self.my_board, &mut self.enemy_board);
        self.is_first = !self.is_first;

        if !self.is_done() && self.legal_actions().len() == 0 {
            self.status = Status::DRAW;
        }
    }
}

impl ConnectFourState {
    fn place_stone(&mut self, action: usize) -> Stone {
        for y in 0..self.params.height {
            if !self.my_board[y][action] && !self.enemy_board[y][action] {
                self.my_board[y][action] = true;
                return Stone { x: action, y };
            }
        }
        panic!("no stone to place");
    }

    fn check_connection(&mut self, first_stone: &Stone, dx: D, dy: D) -> bool {
        let mut que = VecDeque::new();
        que.push_back(first_stone.clone());
        let mut check =
            vec![vec![false; self.params.width]; self.params.height];
        let mut count = 0;
        while !que.is_empty() {
            let stone = que.pop_front().unwrap();
            count += 1;
            if count >= 4 {
                self.status = Status::LOSE;
                return true;
            }
            check[stone.y][stone.x] = true;

            for action in 0..2 {
                let ty = stone.y as isize + dy.value()[action];
                let tx = stone.x as isize + dx.value()[action];
                if ty >= 0
                    && (ty as usize) < self.params.height
                    && tx >= 0
                    && (tx as usize) < self.params.width
                    && self.my_board[ty as usize][tx as usize]
                    && !check[ty as usize][tx as usize]
                {
                    que.push_back(Stone {
                        x: tx as usize,
                        y: ty as usize,
                    });
                }
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ch05::mcts::{mcts_action_arc, MCTSParams};

    fn setup() -> ConnectFourState {
        let params = MazeParams {
            height: 2,
            width: 4,
            end_turn: 0,
        };
        ConnectFourState::new(0, params)
    }

    #[test]
    fn test_use_ch05() {
        let state = setup();
        let params = MCTSParams {
            c: 1.0,
            expand_threshold: 3,
        };
        let actual = mcts_action_arc(10, params)(&state);
        let legal_actions = state.legal_actions();
        assert!(legal_actions.contains(&actual));
    }

    #[test]
    fn test_advance_to_end_game() {
        let mut state = setup();
        state.advance(0);
        state.advance(0);
        state.advance(1);
        state.advance(1);
        state.advance(2);
        state.advance(2);
        assert_eq!(state.is_done(), false);
        state.advance(3);
        assert_eq!(state.is_done(), true);
        assert_eq!(state.white_score(), 1.0);
        assert_eq!(state.teban_score(), 0.0);

        let actual = state.to_string();
        let expected = "\
is_first: false

OOO.
XXXX
";
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_advance() {
        let mut state = setup();
        state.advance(0);
        assert_eq!(state.is_first, false);
        assert_eq!(state.my_board, vec![vec![false; 4]; 2]);
        assert_eq!(
            state.enemy_board,
            vec![vec![true, false, false, false], vec![false; 4]]
        );
    }

    #[test]
    fn test_is_done() {
        let mut state = setup();
        assert_eq!(state.is_done(), false);
        state.status = Status::DRAW;
        assert_eq!(state.is_done(), true);
    }

    #[test]
    fn test_connection() {
        let mut state = setup();
        let stone = Stone { x: 0, y: 0 };
        assert_eq!(state.check_connection(&stone, D::UP, D::STAY), false);

        state.my_board = vec![vec![true; 4], vec![false; 4]];
        assert_eq!(state.check_connection(&stone, D::UP, D::STAY), true);
    }

    #[test]
    fn test_white_score() {
        let mut state = setup();
        state.status = Status::WIN;
        state.is_first = false;
        assert_eq!(state.white_score(), 0.0);
    }

    #[test]
    fn test_teban_score() {
        let mut state = setup();
        state.status = Status::WIN;
        assert_eq!(state.teban_score(), 1.0);
    }

    #[test]
    fn test_to_string() {
        let state = setup();
        let actual = state.to_string();
        let expected = "\
is_first: true

....
....
";
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_place_stone() {
        let mut state = setup();
        let actual = state.place_stone(0);
        let expected = Stone { x: 0, y: 0 };
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_legal_actions() {
        let state = setup();
        let actual = state.legal_actions();
        let expected = vec![0, 1, 2, 3];
        assert_eq!(actual, expected);
    }

    #[test]
    fn make_state() {
        let state = setup();
        assert_eq!(state.params.width, 4);
    }
}
